File: PART 2/Chapter 05/Problem 03/4.py
from collections import deque

N = int(input())
dq = deque()
dq.append((N, 0))
chk = [False] * (N + 1)
chk[N] = True

while dq:
    X, d = dq.popleft()
    if X == 1:
        print(d)
        break

    if X % 3 == 0 and not chk[X // 3]:
        dq.append((X // 3, d + 1))
        chk[X // 3] = True

    if X % 2 == 0 and not chk[X // 2]:
        dq.append((X // 2, d + 1))
        chk[X // 2] = True

    if not chk[X - 1]:
        dq.append((X - 1, d + 1))
        chk[X - 1] = True

File: PART 2/Chapter 01/Problem 02/4.swift
import Foundation

struct Stack<T> { // struct, class doesn't matter
    var arr = [T]()
    
    /// O(1)
    mutating func push(_ val: T) {
      arr.append(val)
    }

    /// O(1)
    mutating func pop() -> T? {
      if arr.isEmpty {
        return nil
      }
      return arr.removeFirst()
    }
    
    /// O(1)
    func peek() -> T? {
      arr.last
    }
}

let tot = Int(readLine() ?? "") ?? 0
var chk = [[String]]()

for _ in 0 ..< tot {
    chk.append(
        (readLine() ?? "")
            .map({ String($0) }) // String aka [Character] -> [String]
    )
}

var stk = Stack<String>()

func chkVPS(_ input: [String]) -> Bool {
    defer {
      stk.arr.removeAll()
    }

    for ch in input {
        if ch == "(" {
            stk.push(ch)
        } else {
            if stk.pop() == nil {
                return false
            }
        }
    }
    
    return stk.peek() != nil ? false : true
}

for val in chk {
  print(chkVPS(val) ? "YES" : "NO")
}